mdata/aggmetric.go

```diff
@@ -54,99 +58,15 @@ func NewAggMetric(store Store, key string, chunkSpan, numChunks uint32, ttl uint
 func (a *AggMetric) SyncChunkSaveState(ts uint32) {
 	a.Lock()
 	defer a.Unlock()
-	chunk := a.getChunkByT0(ts)
-	if chunk != nil {
-		if LogLevel < 2 {
-			log.Debug("AM marking chunk %s:%d as saved.", a.Key, chunk.T0)
-		}
-		chunk.Saved = true
-	}
-}
-
-/* Get a chunk by its T0.  It is expected that the caller has acquired a.Lock()*/
-func (a *AggMetric) getChunkByT0(ts uint32) *chunk.Chunk {
-	// we have no chunks.
-	if len(a.Chunks) == 0 {
-		return nil
-	}
-
-	currentT0 := a.Chunks[a.CurrentChunkPos].T0
-
-	if ts == currentT0 {
-		//found our chunk.
-		return a.Chunks[a.CurrentChunkPos]
-	}
-
-	// requested Chunk is not in our dataset.
-	if ts > currentT0 {
-		return nil
-	}
-
-	// requested Chunk is not in our dataset.
-	if len(a.Chunks) == 1 {
-		return nil
-	}
-
-	// calculate the number of chunks ago our requested T0 is,
-	// assuming that chunks are sequential.
-	chunksAgo := int((currentT0 - ts) / a.ChunkSpan)
-
-	numChunks := len(a.Chunks)
-	oldestPos := a.CurrentChunkPos + 1
-	if oldestPos >= numChunks {
-		oldestPos = 0
+	if ts > a.lastSaveFinish {
+		a.lastSaveFinish = ts
 	}
-
-	var guess int
-
-	if chunksAgo >= (numChunks - 1) {
-		// set guess to the oldest chunk.
-		guess = oldestPos
-	} else {
-		guess = a.CurrentChunkPos - chunksAgo
-		if guess < 0 {
-			guess += numChunks
-		}
+	if ts > a.lastSaveStart {
```

Comment:
taking that a bit further. the CassandraStore guarantees that on a per aggmetric-level (actually per ChunkWriteRequest but CWR's keys is simply the key of the aggmetric), all chunk writes are in order, so the ts will be in order.

also I believe that persist has an order guarantee: the chunks it persists are in ascending time order, both within 1 call, and across multiple calls.

I like the SaveStart and SaveFinish terminology, and i think we should apply it broader.  so instead of `SyncChunkSaveState`, how about this:

```
// saveStart lets us mark up to which chunk t0 we've triggered persistence operations. assumes a is locked
func (a *AggMetric) saveStart(ts uint32) {
    a.lastSaveStart = ts
}
// SaveFinish lets external store packages mark up to which chunk t0 has been saved
func (a *AggMetric) SaveFinisht(ts uint32) {
	a.Lock()
	defer a.Unlock()
        a.lastSaveStart = ts
}
```

the former would be called by persist, the latter by cassandraStore.